File: Chapter 04/Python/section_4_4_anti_patterns/after_rich_domain_thin_controller/infrastructure/repositories.py
from ..domain.interfaces.data_access_interfaces import IOrderRepository, ICustomerRepository, IItemRepository
from ..domain.models.customer import Customer
from ..domain.models.item import Item
class SqlOrderRepository(IOrderRepository):
    """
    ARCHITECTURE NOTE: By isolating SQL logic here, we prevent 
    database concerns from "leaking" into the Presentation or 
    Business layers.
    """
    # Concrete implementation for a SQL database (simulated)
    def get_by_id(self, order_id: int):
        # In a real app, this would perform a SQL query
        print(f"(INFRA) SQL: Fetching Order {order_id}")
        return None
        
    def save(self, order) -> None:
        print(f"(INFRA) SQL: Saving Order {order.id} with Total {order.total_price}")

class SqlCustomerRepository(ICustomerRepository):
    """Concrete implementation for a SQL database (Simulated)"""
    def get_by_id(self, customer_id: int) -> Customer:
       print(f"(INFRA) SQL: Fetching Customer {customer_id}")
        # Returning a dummy Gold customer to test the Rich Domain logic
       return Customer(id=customer_id, type="Gold", email="gold@example.com")
    
class SqlItemRepository(IItemRepository):
    """
    CONCRETE INFRASTRUCTURE IMPLEMENTATION
    ARCHITECTURE NOTE: This is the "Security Guard" of our system. 
    By fetching the Item directly from the database here, we ensure 
    the 'Price' used in calculations is the official one, not a 
    manipulated value sent from a malicious user's JSON request.
    """
    def get_by_id(self, item_id: int) -> Item:
        print(f"(INFRA) SQL: Fetching official price for Item {item_id}")
        
        # Simulated database lookup
        if item_id == 1:
            return Item(id=1, price=100.0, quantity=0)
        elif item_id == 2:
            return Item(id=2, price=50.0, quantity=0)  
        return None

File: Chapter 04/JS/section_4_4_anti_patterns/after_rich_domain_thin_controller/infrastructure/repositories.js
const { 
    IOrderRepository, 
    ICustomerRepository, 
    IItemRepository, 
    IEmailService 
} = require('../domain/interfaces/dataAccessInterfaces');
const Item = require('../domain/models/item');
const Customer = require('../domain/models/customer');
/**
 * ARCHITECTURE NOTE: By isolating Email logic here, we prevent 
 * database concerns from "leaking" into the Presentation or 
 * Business layers.
 * * @implements {IEmailService}
 */
class SmtpEmailService extends IEmailService {
    // Concrete implementation for an email provider
    send(to, subject, body) {
        // Implementation logic would go here
    }
}
class SqlOrderRepository extends IOrderRepository {
    save(order) {
        console.log(`  [DB] SQL: Saving Order ${order.id} with Total ${order.totalPrice.toFixed(2)}`);
    }
}

class SqlCustomerRepository extends ICustomerRepository {
    getById(customerId) {
        console.log(`  [DB] SQL: Fetching Customer ${customerId}`);
        // Returning a dummy Gold customer to trigger the Rich Domain logic
        return new Customer(customerId, "Gold", "gold@example.com");
    }
}

/**
 * INFRASTRUCTURE LAYER: SQL IMPLEMENTATION
 * Simulates a database lookup to ensure we get the official, secure price.
 * * @implements {IItemRepository}
 */
class SqlItemRepository extends IItemRepository {
    getById(itemId) {
        console.log(`  [DB] Fetching official price for Item ID: ${itemId} from SQL.`);

        // Simulated database lookup
        if (itemId === 1) {
            return new Item(1, 100.0, 0);
        } else if (itemId === 2) {
            return new Item(2, 50.0, 0);
        }
        return null;
    }
}

module.exports = { 
    SqlOrderRepository, 
    SqlCustomerRepository, 
    SqlItemRepository, 
    SmtpEmailService 
};